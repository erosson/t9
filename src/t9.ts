/**
 * a single digit of a T9 word. example: `4` (representing "h")
 */
export type Digit = number;

/**
 * a T9-encoded word; the user input. example: `"43556"` (representing "hello")
 */
export type Word = string;

/**
 * given a T9-encoded word, find some matching dictionary words
 */
export type Index = { [code: Word]: IndexEntry };

/**
 * Matching dictionary words for a T9 string
 */
type IndexEntry = {
  /**
   * Commonly-used exact matches for this input, ordered by popularity. Rank these first, in order
   */
  popular: IndexEntrySource;

  /**
   * Uncommonly-used exact matches for this input, unordered. Rank these last
   */
  dict: IndexEntrySource;
};
type IndexEntrySource = {
  /**
   * Exact matches
   */
  words: string[];
  /**
   * This input's a prefix of these other inputs, suggest them too
   */
  prefix: Word[];
};
function emptySource(): IndexEntrySource {
  return { words: [], prefix: [] };
}

export function buildIndex(source: {
  popular: Iterable<string>;
  dict: Iterable<string>;
}): Index {
  const popular = buildSource(source.popular, (s, w) => s !== "" && w !== "");
  const dict = buildSource(
    source.dict,
    (s, w) => s !== "" && w !== "" && !popular[w]?.words?.length
  );
  const index: Index = {};
  // merge sources, grouping by word
  const keys = Object.keys(popular).concat(Object.keys(dict));
  for (let key of Array.from(keys)) {
    if (!(key in index)) {
      index[key] = {
        popular: popular[key] ?? emptySource(),
        dict: dict[key] ?? emptySource(),
      };
    }
  }
  return index;
}
function buildSource(
  words: Iterable<string>,
  filter: (s: string, w: Word) => boolean
): { [w: Word]: IndexEntrySource } {
  const index: { [w: Word]: IndexEntrySource } = {};
  for (let word of words) {
    let t9 = fromWord(word);
    if (filter(word, t9)) {
      index[t9] = index[t9] || emptySource();
      // no dupes
      if (index[t9].words.findIndex((w) => w === word) < 0) {
        index[t9].words.push(word);
      }

      for (let prefix of prefixes(t9)) {
        index[prefix] = index[prefix] || emptySource();
        index[prefix].prefix.push(t9);
      }
    }
  }
  return index;
}

function prefixes(word: Word): Word[] {
  return range(word.length - 1).map((i) => word.slice(0, i));
}
function range(n: number): number[] {
  return Array.from(Array(n).keys());
}

/**
 * map digits to the letters each represents. `{2: "abc", 3: "def", ...}`
 */
export const letters = [
  "",
  "",
  "abc",
  "def",
  "ghi",
  "jkl",
  "mno",
  "pqrs",
  "tuv",
  "wxyz",
];

/**
 * inverted `letters` table; map letters to their digits.
 * `{a: 2, b: 2, c: 2, d: 3, e: 3, f: 3, ...}`
 */
export const byLetter: { [l: string]: number } = Object.fromEntries(
  letters.flatMap((ls, n) => ls.split("").map((l) => [l, n]))
);

/**
 * transform a string to a T9 word. `"Hello!"` => `"43556"`
 */
export function fromWord(word: string): Word {
  return word
    .toLowerCase()
    .split("")
    .map((l) => byLetter[l])
    .filter((t9) => !!t9)
    .join("");
}

type Result = { word: string; source: string };

/**
 * Given an index and some input, rank the possible words.
 */
export function ranking(index: Index, input: Word): Result[] {
  if (!input) return [];
  const numeric: Result = { word: input, source: "numeric" };
  const data = index[input];
  if (!data) return [numeric];
  const popularPrefix = data.popular.prefix.flatMap(
    (p) => index[p].popular.words
  );
  const dictPrefix = data.dict.prefix.flatMap((p) =>
    p.length >= 3 ? index[p].dict.words : []
  );
  return [
    data.popular.words.map((word) => ({ word, source: "popular-words" })),
    popularPrefix.map((word) => ({ word, source: "popular-prefix" })),
    data.dict.words.map((word) => ({ word, source: "dict-words" })),
    dictPrefix.map((word) => ({ word, source: "dict-prefix" })),
    numeric,
  ].flat();
}

export function words(index: Index, input: Word): string[] {
  return ranking(index, input).map((r) => r.word);
}
