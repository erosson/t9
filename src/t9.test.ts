import * as T9 from "./t9";
import * as fs from "fs/promises";

const data = [
  { input: "43556", output: ["hello", "Hellman", "Hellman's", "43556"] },
  {
    input: "463",
    output: [
      "god",
      "inf",
      "ind",
      "imf",
      "gnd",
      "information",
      "index",
      "industry",
      "individual",
      "india",
    ],
  },
  {
    input: "4636",
    output: [
      "info",
      "indo",
      "information",
      "informed",
      "indonesia",
      "indoor",
      "inform",
      "informational",
      "informal",
      "informative",
    ],
  },
  {
    input: "6666",
    output: [
      "moon",
      "noon",
      "mono",
      "monopoly",
      "monogram",
      "moonlight",
      "monoclonal",
      "monophonic",
      "monmouth",
      "Monmouth's",
    ],
  },
  {
    input: "666666666666",
    output: ["666666666666"],
  },
];

let index: T9.Index;

beforeAll(async () => {
  const [dict, popular] = await Promise.all([
    fs
      .readFile(`${__dirname}/../data/words.txt`)
      .then((f) => f.toString().split("\n")),
    fs
      .readFile(`${__dirname}/../data/20k.txt`)
      .then((f) => f.toString().split("\n")),
  ]);
  index = T9.buildIndex({ dict, popular });
});

test.each(data)("test input: $input", ({ input, output }) => {
  expect(T9.words(index, input).slice(0, 10)).toEqual(output);
  expect(T9.ranking(index, input).slice(0, 10)).toMatchSnapshot(input);
});
