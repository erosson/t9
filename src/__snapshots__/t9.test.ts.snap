// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test input: 463: 463 1`] = `
[
  {
    "source": "popular-words",
    "word": "god",
  },
  {
    "source": "popular-words",
    "word": "inf",
  },
  {
    "source": "popular-words",
    "word": "ind",
  },
  {
    "source": "popular-words",
    "word": "imf",
  },
  {
    "source": "popular-words",
    "word": "gnd",
  },
  {
    "source": "popular-prefix",
    "word": "information",
  },
  {
    "source": "popular-prefix",
    "word": "index",
  },
  {
    "source": "popular-prefix",
    "word": "industry",
  },
  {
    "source": "popular-prefix",
    "word": "individual",
  },
  {
    "source": "popular-prefix",
    "word": "india",
  },
]
`;

exports[`test input: 4636: 4636 1`] = `
[
  {
    "source": "popular-words",
    "word": "info",
  },
  {
    "source": "popular-words",
    "word": "indo",
  },
  {
    "source": "popular-prefix",
    "word": "information",
  },
  {
    "source": "popular-prefix",
    "word": "informed",
  },
  {
    "source": "popular-prefix",
    "word": "indonesia",
  },
  {
    "source": "popular-prefix",
    "word": "indoor",
  },
  {
    "source": "popular-prefix",
    "word": "inform",
  },
  {
    "source": "popular-prefix",
    "word": "informational",
  },
  {
    "source": "popular-prefix",
    "word": "informal",
  },
  {
    "source": "popular-prefix",
    "word": "informative",
  },
]
`;

exports[`test input: 6666: 6666 1`] = `
[
  {
    "source": "popular-words",
    "word": "moon",
  },
  {
    "source": "popular-words",
    "word": "noon",
  },
  {
    "source": "popular-words",
    "word": "mono",
  },
  {
    "source": "popular-prefix",
    "word": "monopoly",
  },
  {
    "source": "popular-prefix",
    "word": "monogram",
  },
  {
    "source": "popular-prefix",
    "word": "moonlight",
  },
  {
    "source": "popular-prefix",
    "word": "monoclonal",
  },
  {
    "source": "popular-prefix",
    "word": "monophonic",
  },
  {
    "source": "popular-prefix",
    "word": "monmouth",
  },
  {
    "source": "dict-prefix",
    "word": "Monmouth's",
  },
]
`;

exports[`test input: 43556: 43556 1`] = `
[
  {
    "source": "popular-words",
    "word": "hello",
  },
  {
    "source": "dict-prefix",
    "word": "Hellman",
  },
  {
    "source": "dict-prefix",
    "word": "Hellman's",
  },
  {
    "source": "numeric",
    "word": "43556",
  },
]
`;

exports[`test input: 666666666666: 666666666666 1`] = `
[
  {
    "source": "numeric",
    "word": "666666666666",
  },
]
`;
